src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -280,6 +308,257 @@ size_t runtime::gpu::CUDAEmitter::build_1d_max_pool(const GPURuntimeContext* ctx
+        shape_params[T_] = 1;
+        shape_params[R_] = 1;
+        shape_params[S_] = static_cast<int>(window[0]);
+        shape_params[STRIDE_D] = 0;
+        shape_params[STRIDE_H] = 0;
+        shape_params[STRIDE_W] = static_cast<int>(strides[0]);
+        shape_params[PAD_D] = 0;
+        shape_params[PAD_H] = 0;
+        shape_params[PAD_W] = static_cast<int>(pad[0]);
+    }
+    else if (in.size() == 4)
+    {
+        shape_params[D_] = 1;
+        shape_params[H_] = static_cast<int>(in[2]);
+        shape_params[W_] = static_cast<int>(in[3]);
+        shape_params[M_] = 1;
+        shape_params[P_] = static_cast<int>(out[2]);
+        shape_params[Q_] = static_cast<int>(out[3]);
+        shape_params[T_] = 1;
+        shape_params[R_] = static_cast<int>(window[0]);
+        shape_params[S_] = static_cast<int>(window[1]);
+        shape_params[STRIDE_D] = 0;
+        shape_params[STRIDE_H] = static_cast<int>(strides[0]);
+        shape_params[STRIDE_W] = static_cast<int>(strides[1]);
+        shape_params[PAD_D] = 0;
+        shape_params[PAD_H] = static_cast<int>(pad[0]);
+        shape_params[PAD_W] = static_cast<int>(pad[1]);
+    }
+    else if (in.size() == 5)
+    {
+        shape_params[D_] = static_cast<int>(in[2]);
+        shape_params[H_] = static_cast<int>(in[3]);
+        shape_params[W_] = static_cast<int>(in[4]);
+        shape_params[M_] = static_cast<int>(out[2]);
+        shape_params[P_] = static_cast<int>(out[3]);
+        shape_params[Q_] = static_cast<int>(out[4]);
+        shape_params[T_] = static_cast<int>(window[0]);
+        shape_params[R_] = static_cast<int>(window[1]);
+        shape_params[S_] = static_cast<int>(window[2]);
+        shape_params[STRIDE_D] = static_cast<int>(strides[0]);
+        shape_params[STRIDE_H] = static_cast<int>(strides[1]);
+        shape_params[STRIDE_W] = static_cast<int>(strides[2]);
+        shape_params[PAD_D] = static_cast<int>(pad[0]);
+        shape_params[PAD_H] = static_cast<int>(pad[1]);
+        shape_params[PAD_W] = static_cast<int>(pad[2]);
+    }
+    else
+    {
+        throw std::runtime_error("AvgPool currently supports up to 3 spatial dimensions.");
+    }
+    return shape_params;
+}
+
+size_t runtime::gpu::CUDAEmitter::build_avg_pool(const GPURuntimeContext* ctx,
+                                                 const std::array<std::string, 2>& dtypes,
+                                                 const Shape& input_shape,
+                                                 const Shape& output_shape,
+                                                 const Shape& window_shape,
+                                                 const Shape& window_stride,
+                                                 const Shape& padding_below,
+                                                 bool include_pad)
+{
+    // assumes NCDHW format
+    auto input_width = input_shape.back();
+    auto output_width = output_shape.back();
+    auto s =
```

Comment:
If you are calling a function `get_avgpool_params` then name the variable `avgpool_params` if for no other reason than to make it easy to search for in the code. It took me a while to find it's use because I can't search for `s`.
I think this is not used until way later in the code. Why not move this line down to be closer to where it is ultimately used.

Reply:
Agreed, I think it is more clear this way, I replaced the inline enum with a struct as you suggest. However, the pooling shape parameters are used in the function hash of the kernel to determine if the kernel has been used previously. This is why this line appears early on. I also updated the name as suggested.